index.js

```diff
@@ -63,11 +51,145 @@ async function gatherResponse(response) {
+  if (!bodydata.address || !emailRegex.test(bodydata.address)) return false
+
+  return true
+}
+
 async function handleRequest(request) {
-  const formData = await request.formData()
-  const bodydata = {}
-  for (const entry of formData.entries()) {
-    bodydata[entry[0]] = entry[1]
+
+  try
+  {
+    //
+    // \TODO !! Uncomment the origin checking for the live version !!
+    //
+    // Validate the origin of the request
+    if( request.method === 'POST' ) // && request.url.hostname === '...' && request.url.pathname === '/add')
+    {
+      console.log("handling");
+
+      const formData = await request.formData()
+      const bodydata = {}
+      const listsToSubscribeTo = [] // Empty array to hold all the email lists to sign up to
+      for (const entry of formData.entries()) {
+        bodydata[entry[0]] = entry[1]
+
+        // Add the requested email lists to the array
+        if( entry[0] === "signup-checkbox" ) {
+          listsToSubscribeTo.push(entry[1]);
+        }
+      }
+
+      // Toggle subscribed
+      bodydata['subscribed'] = true
+      // Update user if present
+      bodydata['upsert'] = true      
+
+      // Validate the submitted data
+      if (!validateFormData(bodydata)) {
+        return new Response('Invalid submission', { status: 400, headers:corsHeaders })
+      }
+
+      // Extract the recaptcha token
+      const recaptchaToken = bodydata['g-recaptcha-response']
+      if (!recaptchaToken) {
+        return new Response('Invalid reCAPTCHA', { status: 400, headers:corsHeaders })
+      }
+
+      const recaptchaResponse = await fetch(
+        `https://www.google.com/recaptcha/api/siteverify?secret=${RECAPTCHASECRET}&response=${recaptchaToken}`, {
+          method: 'POST'
+      })
+
+      const recaptchaBody = await recaptchaResponse.json()
+      if (recaptchaBody.success == false) {
+        return new Response('reCAPTCHA failed', { status: 400, headers:corsHeaders })
+      }
+
+      // At this point, we passed the captcha and we have valid entries
+      let bodyoptions = {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/x-www-form-urlencoded",
+          "Content-Length": bodydata.length,
+          'Authorization': 'Basic ' + base64encodedData,
+        },
+        body: urlfy(bodydata),
+      }
+
+      for( var i = 0; i < listsToSubscribeTo.length; i++ ) {
+        var mailgunListName = "";
+        switch( listsToSubscribeTo[i] ) {
+          case "signup-general":
+            mailgunListName = "vsf-announce"
+            break;
+          case "signup-speakerscorner":
+            mailgunListName = "speakers_corner"
+            break;
+          default:
+            return new Response("${mailgunListName} cannot be subscribed to via this URL", {status:403, headers:corsHeaders})
```

Comment:
Did you mean to use backticks here? Otherwise it isn't a template string.

Reply:
Yes